feat(medusa): Allow to assign a collection to a product on import

adrien2p opened the pull request:
**What**
It is not possible to assign a product to a collection through the import feature, even though we support passing a collection in the data to import.

This PR adds support for collection to be handled as part of the import process.

As per the discussion on linear, the collections must be created prior the import.

**Tests**
The integration tests have been updated to validate that the products are associated with the right collection

FIXES CORE-900

fPolic commented on packages/medusa/src/strategies/batch-jobs/product/import.ts:
```diff
@@ -463,6 +486,53 @@ class ProductImportStrategy extends AbstractBatchJobStrategy {
+        throw e
+      })
+
+    if (shouldCreateCollection) {
```
context: previous discussion about this - https://linear.app/medusajs/issue/CORE-700#comment-59937c68

adrien2p replied:
Alright, any reason for that? The reason is not mentioned in the ticket

adrien2p replied:
I have removed that part 🎉

RegisHubelia commented:
Testing this now - I have a pretty big test file too (32k products) - so I'll let you know how it goes and get back to you.

adrien2p commented:
> Testing this now - I have a pretty big test file too (32k products) - so I'll let you know how it goes and get back to you.

Just a sec, I am updating it, I ll let you know

olivermrbl commented on packages/medusa/src/strategies/batch-jobs/product/import.ts:
```diff
@@ -463,6 +486,42 @@ class ProductImportStrategy extends AbstractBatchJobStrategy {
 
+  private async assignProductToCollection(
+    product: Product,
+    collectionData: { title: string; handle: string }
```
comment: We don't use title for anything, so we should we change the method signature to be `assignProductToCollection(product, collectionHandle)`?

adrien2p replied:
I ve changed the implementation, this method does not exists anymore. I am testing before pushing

adrien2p commented:
> Testing this now - I have a pretty big test file too (32k products) - so I'll let you know how it goes and get back to you.

@RegisHubelia I suggest you to wait for the pipeline to be green and then you can test.

RegisHubelia commented:
K!

pKorsholm commented on packages/medusa/src/strategies/batch-jobs/product/import.ts:
```diff
@@ -405,6 +411,16 @@ class ProductImportStrategy extends AbstractBatchJobStrategy {
+          productOp["product.collection.handle"] != null &&
+          productOp["product.collection.handle"] !== ""
+        ) {
+          productData["collection"] =
+            await productCollectionServiceTx.retrieveByHandle(
+              productOp["product.collection.handle"] as string
+            )
```
nit: `collection` is not a property of `CreateProductInput`, we have used `collection_id` until now. We should either allow `Collection` to be passed on "dot" the id from the result of `retrieveByHandle`.

pKorsholm commented on packages/medusa/src/strategies/batch-jobs/product/import.ts:
```diff
@@ -451,6 +469,16 @@ class ProductImportStrategy extends AbstractBatchJobStrategy {
+          productOp["product.collection.handle"] != null &&
+          productOp["product.collection.handle"] !== ""
+        ) {
+          productData["collection"] =
+            await productCollectionServiceTx.retrieveByHandle(
+              productOp["product.collection.handle"] as string
+            )
+        }
```
same as above

pKorsholm reviewed (commented):
very neat! I have one comment regarding the typing of create and update inputs but otherwise it looks good! 😄

adrien2p commented:
@olivermrbl @fPolic @pKorsholm I have updated the way we assign the collection to the product. 
With the actual production import, we were able to attach a collection only if the title and handle was matching an existing collection. The problem is that the `productData` has been cast to what the product service methods were expecting to receive even though they are not at all and are different by nature. Therefore everything coming from the csv was passed down to the service including the collection. 
Since it is not part of that PR to clean up that. I have just assigned the collection id to the data and deleted the collection, that way there is no conflict between the props whenTypeorm handles them.

olivermrbl reviewed (commented):
~~LGTM!~~

Tests need fixing

@RegisHubelia let us know when you've had a chance to test it out - curious to hear the result :)

RegisHubelia commented:
I tested the last pr - will test the latest changes in a few but worked fine - was going to talk about the title vs handle but I see that you covered it.  Waiting for the pipeline to finish, will pull and check if all good and get back asap - hopefully we can release this quickly

olivermrbl commented on integration-tests/api/factories/simple-product-collection-factory.ts:
```diff
@@ -0,0 +1,39 @@
+    const collection_ = manager.create(ProductCollection, {
+      id: `simple-id-${Math.random() * 1000}`,
+      title: collectionData.title ?? faker.datatype.string(10),
+      handle: collectionData.handle
```
comment: Sorry, I actually meant to use it for `title` and omit handle as we have a `beforeInsert` trigger, that kebab-cases the title. That way, we don't have to require any data for the factory.

But really, you don't have to include it. It's nitpicking 😅

adrien2p replied:
I can make the handle optional but i will still have to pass some data to the factory as I am dependent from the data in the csv template for the tests ^^

adrien2p replied:
I think we can merge as it is for now wdyt?

olivermrbl replied:
Yep!

RegisHubelia replied:
The problem I see with using the title is that you can have typos...  ie: in our use case, we had a csv, and some collection title had for example "A Collection Title", and some other had "A-Collection Title"  and another "A Collections Title" but the handle is the same...  It's easier not to messup with fewer products, but in our case there are 32k products...  Not the end of the world tough, we can go back and manually reassign.

adrien2p replied:
@RegisHubelia this is a factory used for the integration tests and it is not related to the strategy implementation. This is also why the strategy is based on the handle and not the title 💪

adrien2p replied:
@olivermrbl I have included it in the end :D

patrick-medusajs reviewed (approved):
LGTM 💪

RegisHubelia commented:
Works!  Tried with 1k product and everything went smoothly!

RegisHubelia commented:
Hey Guys, I found an issue with the exports...  On 1.6.5, when I export, the size of the file is consistently 79.91M with 22695 rows.  Using the feature branch, it's inconsistent.  On 3 different export, size varied from 78.90M with 22544 rows, 79.8M with 22683 rows, and 82.5M with 22956 rows.  Basically, each time I export on the feature branch, I get a different file size and rows...  I see some changes in the export file, and I think this is where the inconsistency comes from.  But for sure something is up with exporting with this PR...

adrien2p commented:
> Hey Guys, I found an issue with the exports... On 1.6.5, when I export, the size of the file is consistently 79.91M with 22695 rows. Using the feature branch, it's inconsistent. On 3 different export, size varied from 78.90M with 22544 rows, 79.8M with 22683 rows, and 82.5M with 22956 rows. Basically, each time I export on the feature branch, I get a different file size and rows... I see some changes in the export file, and I think this is where the inconsistency comes from. But for sure something is up with exporting with this PR...

This behaviour looks pretty weird. Nothing have been changed in the export except from a code formatting POV [here](https://github.com/medusajs/medusa/pull/2764/files#diff-8f0252a01b6052bb9c56e7739274e0dc411dd038915eb0ee077d69777e369436R507)

adrien2p commented:
@RegisHubelia Would you be able to share the content of the batch jobs [context, result] columns please

RegisHubelia commented:
@adrien2p, can I send you a db dump?  This is client data so I cannot really share it here...  I switched to the develop branch, and the exports seems consistent now, but re-importing the exact same exported file without any modification (didn't even opened the file) fails saying the Product Handle column is missing.  I'd also send you a file where we added the collections (that's what I've been trying to do for the last few days without much success with larger files).  Very strange...

adrien2p commented:
@RegisHubelia when you export the data, in the file there should be a column `Product Handle`, when I check at the export it should appear, could you confirm?

RegisHubelia commented:
Yes the column is there...  I think there is a delimiter/csv format issue...  The delimiter in medusa is ';', this caused me a bit of a headhache as the products imported from shopify had css and html in the description, so we ended up removing those and eventually get the import to work, but it's really been a hit and miss chase...  I'll send the file to your email.

adrien2p commented:
When we export the data, we wrap the cell content into `"` if there is a new line. I will try to do the same if there is a coma or semicolon. Unfortunately, separator are always a pain if the platefor that export the file does not wrap the data between `"` which we do for content with new line.

RegisHubelia commented:
Well actually, maybe the best way to deal with this would be to wrap all cells into "?  Then again we can end up with " in the description or other, and this is where it becomes annoying I guess...  I agree, csv is always a pain...  Maybe JSON export/imports..?  Not as easy to edit etc, but it would probably be less painful when weird chars are involved...

adrien2p commented:
Our csv cell formatter already handle if there is a " and will escape it but only if the cell need to be wrapped as it does not need to be done otherwise.

adrien2p commented:
Yes I am adding that, also it will prevent any issues in that matter. The export wrap and the import unwrap

adrien2p commented:
Also, the file to import should be splitting the data into chunks because otherwise it is consuming a lot of memory on the machine and can slow down the entire core process or even block them during the processing. what do you think @olivermrbl

adrien2p commented:
It is running, so far so good 👍

RegisHubelia commented:
Perfect, thanks.  I have a dev environment so no worries here.  Thanks a bunch!!

adrien2p commented:
could you test that pr @RegisHubelia https://github.com/medusajs/medusa/pull/2789

RegisHubelia commented:
Doing this now!
